### PR TITLE
Tests for LateralProfile

### PR DESCRIPTION
**Pull request by MarekOchocki**

### CODE REVIEW

**Comment by grzanka** [comment]
@MarekOchocki investigate why tests are failing

**Comment by MarekOchocki** [comment]
Tests are failing, because of issue #15

**Comment by grzanka** [comment]
I agree with that, but what about this problem:
```
>       self.assertEqual(self.data2.penumbra_right(), 0.64)
E       AssertionError: 0.6400000000000001 != 0.64
```
?

Will you fix issue #15 in a commit belonging to this PR ?

**Comment by MarekOchocki** [comment]
Yes. Last commit in this PR fixes issue #15

File: ibodata/lateral_profile.py [hunk 1]
```diff
@@ -120,7 +120,19 @@ def normalize(self, dt, allow_cast=True):
         mid = self.x_at_y(a) + w / 2.0
         self.x -= mid
 
-        ave = np.average(self.y[np.fabs(self.x) <= dt])
+        slice_y = self.y[np.fabs(self.x) <= dt]
+        slice_x = self.x[np.fabs(self.x) <= dt]
+
+        if np.argwhere(self.x == -dt).size == 0:
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Add a comment explaining why this corner case require special treatment.

File: ibodata/lateral_profile.py [hunk 2]
```diff
@@ -120,7 +120,19 @@ def normalize(self, dt, allow_cast=True):
...
+            slice_y = np.append(slice_y, self.y_at_x(dt))
+            slice_x = np.append(slice_x, dt)
+
+        if np.argwhere(self.x == dt).size == 0:
+            slice_y = np.append(self.y_at_x(-dt), slice_y)
+            slice_x = np.append(-dt, slice_x)
+
+        area = np.trapz(slice_y, slice_x)
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Why first goes y then x ?

Comment (MarekOchocki) [reply ibodata/lateral_profile.py]: That's normal usage of numpy.trapz() function: https://docs.scipy.org/doc/numpy/reference/generated/numpy.trapz.html

File: ibodata/lateral_profile.py [hunk 3]
```diff
@@ -120,7 +120,24 @@ def normalize(self, dt, allow_cast=True):
         mid = self.x_at_y(a) + w / 2.0
         self.x -= mid
 
-        ave = np.average(self.y[np.fabs(self.x) <= dt])
+        slice_y = self.y[np.fabs(self.x) <= dt]
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Slice usually referers to indices, not to subset of numpy array. See i.e.:

```
>>> l = ['a', 'b', 'c', 'd' ]
>>> l[2:3]
['c']
>>> slice_tmp = slice(2,3)
>>> l[slice_tmp]
['c']
```

I'd recommend to rename this slices to sth like `norm_core_x` or `norm_section_x` (and `_y` also).

File: ibodata/lateral_profile.py [hunk 4]
```diff
@@ -120,7 +120,24 @@ def normalize(self, dt, allow_cast=True):
...
+        slice_x = self.x[np.fabs(self.x) <= dt]
+
+        """
+        if there's no point on the edge normalization is not perfectly accurate 
+        and we are normalizing over smaller area than [-dt, +dt]
+        That's why we interpolate points on the edge below.
+        """
+        if np.argwhere(self.x == -dt).size == 0:
+            slice_y = np.append(slice_y, self.y_at_x(dt))
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: This is expensive in terms of memory allocation. I'd avoid it.

Why not calculate first area over usuall slices.

Then do such if statements as here and add missing area, using code like:

```
if left_point_not_found:
  x_coords = (-dt, slice_x[0])
  y_coords = (self.y_at_x(-dt), slice_y[0])
  area += np.trapz(x_coords, y_coords)
```

File: ibodata/lateral_profile.py [hunk 5]
```diff
@@ -120,7 +120,26 @@ def normalize(self, dt, allow_cast=True):
...
+
+        """
+        if there's no point on the edge normalization is not perfectly accurate
+        and we are normalizing over smaller area than [-dt, +dt]
+        That's why we interpolate points on the edge below.
+        """
+        if np.argwhere(self.x == -dt).size == 0:
+            coords_y = (self.y_at_x(-dt), norm_section_y[0])
+            coords_x = (-dt, norm_section_x[0])
+            area += np.trapz(coords_y, coords_x)
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: numpy `trapz` has a counterintuitive way of specifying parameters. I'd expect to call it like: `trapz(x,y)`, not `trapz(y,x)`.
To be more specific please use named arguments and change the order of arguments (which is possible when arguments are named):

```
area += np.trapz(x = coords_x, y = coords_y)
```

This will help future developers

File: ibodata/lateral_profile.py [hunk 6]
```diff
@@ -120,7 +120,26 @@ def normalize(self, dt, allow_cast=True):
...
+        norm_section_x = self.x[np.fabs(self.x) <= dt]
+        area = np.trapz(norm_section_y, norm_section_x)
+
+        """
+        if there's no point on the edge normalization is not perfectly accurate
+        and we are normalizing over smaller area than [-dt, +dt]
+        That's why we interpolate points on the edge below.
+        """
+        if np.argwhere(self.x == -dt).size == 0:
+            coords_y = (self.y_at_x(-dt), norm_section_y[0])
```
Comment (grzanka) [inline ibodata/lateral_profile.py:150]: Is there a test which covers this corner case ?

File: tests/lateral.py [hunk 7]
```diff
@@ -35,12 +35,15 @@ def setUp(self):
...
         self.data3 = LateralProfile([[-1.5, 5], [0.5, 3], [1.5, 1]])
+        self.data4 = LateralProfile([[-3, 5], [-1, 8], [0, 8], [1, 4]], dtype=np.int)
 
     def test_normalize(self):
         with self.assertRaises(ValueError):
             self.data1.normalize(1)
         with self.assertRaises(ValueError):
             self.data3.normalize(1)
+        with self.assertRaises(TypeError):
+            self.data4.normalize(1, False)
```
Comment (grzanka) [inline tests/lateral.py]: Add here named arguments `dt, allow_cast`

File: ibodata/lateral_profile.py (hunk 6 above)
Comment (MarekOchocki) [reply ibodata/lateral_profile.py:150]: Yes, it is. Test coverage is currently 100%

File: ibodata/lateral_profile.py (hunk 5 above)
Comment (MarekOchocki) [reply ibodata/lateral_profile.py]: Done

File: ibodata/lateral_profile.py [hunk 8]
```diff
@@ -87,17 +87,17 @@ def asymmetry(self):
...
-        area_left = np.trapz(self.y[self.x <= 0], self.x[self.x <= 0])
-        area_right = np.trapz(self.y[self.x >= 0], self.x[self.x >= 0])
+        area_left = np.trapz(x=self.x[self.x <= 0], y=self.y[self.x <= 0])
+        area_right = np.trapz(x=self.x[self.x >= 0], y=self.y[self.x >= 0])
 
         if np.argwhere(self.x == 0).size == 0:
             left_index_arr = np.argwhere(self.x < 0)
-            area_left += np.trapz(np.append(self.y[left_index_arr[left_index_arr.size - 1]], self.y_at_x(0)),
-                                  np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]))
+            area_left += np.trapz(x=np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]),
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: python supports negative indexing, last item of a table `t` can be obtained by call to `t[-1]`
It looks that `left_index_arr.size` is not needed

File: ibodata/lateral_profile.py [hunk 9]
```diff
@@ -87,17 +87,17 @@ def asymmetry(self):
...
+                                  y=np.append(self.y[left_index_arr[left_index_arr.size - 1]], self.y_at_x(0)))
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Do we really need to use `np.append` here ? Wouldn't it work with simple two-element tuple ?

File: ibodata/lateral_profile.py [hunk 10]
```diff
@@ -109,18 +109,48 @@ def normalize(self, dt, allow_cast=True):
         In case of corrupted data raises ValueError
         Translate y to bring y.min() to 0 (noise substraction) and then
         normalize to 1 over [-dt, +dt] area from the mid of the profile
+        if allow_cast is set to True, division not in place and casting may occur.
+        If division in place is not possible and allow_cast is False
+        an exception is raised.
         """
+        try:
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Explain in comments the reason of such `try` block. Sth like:

We will try now to divide the data by 1.0. If self.y is float, then nothing happens.
In case its has integers there are two cases:
 - conversion to float will happen blabla
 - integer will stay integers bla bla

File: ibodata/lateral_profile.py [hunk 11]
```diff
@@ -109,18 +109,48 @@ def normalize(self, dt, allow_cast=True):
...
 
-        a = self.y.max() - self.y.min()
-        a /= 2.0
+        a = self.y.max() / 2.0
         w = self.width(a)
         if np.isnan(w):
             raise ValueError("Part of profile is missing.")
         mid = self.x_at_y(a) + w / 2.0
-        self.x -= mid
+        if allow_cast:
```
Comment (grzanka) [inline ibodata/lateral_profile.py]: Add some explanation why sometimes in-place addition/subtraction is used and sometimes not.
Here `mid` can be float - is it the reason ? If yes, say it clearly in comments

**Review by grzanka — APPROVED** [review]
